extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedSpanBuilder.java

```diff
@@ -0,0 +1,215 @@
+  public <T> ExtendedSpanBuilder setAttribute(AttributeKey<T> key, T value) {
+    delegate.setAttribute(key, value);
+    return this;
+  }
+
+  @Override
+  public ExtendedSpanBuilder setAllAttributes(Attributes attributes) {
+    delegate.setAllAttributes(attributes);
+    return this;
+  }
+
+  @Override
+  public ExtendedSpanBuilder setSpanKind(SpanKind spanKind) {
+    delegate.setSpanKind(spanKind);
+    return this;
+  }
+
+  @Override
+  public ExtendedSpanBuilder setStartTimestamp(long startTimestamp, TimeUnit unit) {
+    delegate.setStartTimestamp(startTimestamp, unit);
+    return this;
+  }
+
+  @Override
+  public ExtendedSpanBuilder setStartTimestamp(Instant startTimestamp) {
+    delegate.setStartTimestamp(startTimestamp);
+    return this;
+  }
+
+  /**
+   * Sets the exception handler for {@link #call(SpanCallable)} and {@link #run(SpanRunnable)}
+   *
+   * <p>The exception handler gives you the opportunity to handle the exception in a custom way,
+   * e.g. not marking the span as error, which is the default behavior.
+   *
+   * @param exceptionHandler the exception handler
+   */
+  public ExtendedSpanBuilder setExceptionHandler(BiConsumer<Span, Throwable> exceptionHandler) {
+    this.exceptionHandler = exceptionHandler;
+    return this;
+  }
+
+  /**
+   * Extract a span context from the given carrier and set it as parent of the span for {@link
+   * #call(SpanCallable)} and {@link #run(SpanRunnable)}.
+   *
+   * <p>The span context will be extracted from the <code>carrier</code>, which you usually get from
+   * HTTP headers of the metadata of a message you're processing.
+   *
+   * <p>A typical usage would be: <code>
+   * ExtendedTracer.create(tracer)
+   *     .setSpanKind(SpanKind.SERVER)
+   *     .extractContext(propagators, carrier)
+   *     .run("my-span", () -> { ... });
+   * </code>
+   *
+   * @param propagators provide the propagators from {@link OpenTelemetry#getPropagators()}
+   * @param carrier the string map where to extract the span context from
+   */
+  public ExtendedSpanBuilder extractContext(
+      ContextPropagators propagators, Map<String, String> carrier) {
+    this.extractedContext =
+        ExtendedContextPropagators.extractTextMapPropagationContext(carrier, propagators);
+    return this;
+  }
+
+  @Override
+  public Span startSpan() {
+    return delegate.startSpan();
+  }
+
+  /**
+   * Runs the given {@link SpanCallable} inside of the span created by the given {@link
+   * SpanBuilder}. The span will be ended at the end of the {@link SpanCallable}.
+   *
+   * <p>If an exception is thrown by the {@link SpanCallable}, the span will be marked as error, and
+   * the exception will be recorded.
+   *
+   * @param spanCallable the {@link SpanCallable} to call
+   * @param <T> the type of the result
+   * @param <E> the type of the exception
+   * @return the result of the {@link SpanCallable}
+   */
+  public <T, E extends Throwable> T call(SpanCallable<T, E> spanCallable) throws E {
```

Comment:
I kind of like this approach, but don't like the `setExceptionHandler` method. Generally, I think it would be good to avoid having `ExtendedSpanBuilder` have any state besides the delegate SpanBuilder. 

I recommend:
- Removing setException handler and having it as an argument on the call / run methods.
- Rename `call` / `run` to `startAndCall` / `startAndRun`. This makes it clear that these methods are extended versions of the `start` method. The fact that `call` / `run` starts / ends the span is kind of surprising if you don't look at the javadoc / code.
- Rename `extractContext` to `extractAndSetParent`. Change the implementation to call `setParent` with the extracted context.

Reply:
Are you suggesting not to end the span in `startAndCall`? This would kind of defeat the purpose.